core/optaplanner-constraint-streams/src/main/java/org/optaplanner/constraint/streams/bavet/common/index/EqualsAndComparisonIndexer.java

```diff
@@ -41,8 +40,9 @@ public EqualsAndComparisonIndexer(JoinerType comparisonJoinerType) {
     @Override
     public void put(Object[] indexProperties, Tuple_ tuple, Value_ value) {
         Objects.requireNonNull(value);
-        IndexerKey equalsIndexKey = new IndexerKey(Arrays.copyOfRange(indexProperties, 0, indexProperties.length - 1));
-        Object comparisonIndexProperty = indexProperties[indexProperties.length - 1];
+        int indexPropertyCount = indexProperties.length;
+        IndexerKey equalsIndexKey = new IndexerKey(indexProperties, indexPropertyCount - 1);
```

Comment:
Don't make a copy of the array. Instead `IndexerKey` has a tailored equals/hashCode that only counts up to the given index.
This brings the first half of the massive perf boost.